### PR TITLE
[LTC] Get rid of current device concept

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
The current device concept is used in three scenarios, and each either now
has a better replacement or is not applicable. Here they are:
1) Used by UAPI _ltc_set_default_device/_ltc_get_default_device: these APIs
are available for users before but actually make no senses. It makes sense
to set default device type to CPU, GPU or TPU, but not lazy given that would
just use whatever default hardware type. Setting ordinal like lazy:0 doesn't
make any sense now as distributed training is still under development.
2) Used by LTCGuardImpl: it only needs to know c10 device but not backend
device. Replaced the g_device with c10 one.
3) Used as the default device: replaced with Device(), which later will
be enhanced with the default device type backend API.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

### CODE REVIEW

File: lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp [hunk 1]
```diff
@@ -54,7 +54,7 @@ c10::optional<Device> GetOptionalDevice(const std::string& device_str) {
 
 Device GetDeviceOrCurrent(const std::string& device_str) {
   if (device_str.empty()) {
-    return GetCurrentDevice();
+    return Device();
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp:57]: so what does this do exactly?

File: lazy_tensor_core/lazy_tensor_core/csrc/tensor_impl.cpp [hunk 2]
```diff
@@ -10,31 +10,40 @@
...
   }
 
   c10::Device getDevice() const override {
-    return bridge::GetCurrentAtenDevice();
+    return g_device;
   }
 
   void setDevice(c10::Device device) const override {
-    bridge::SetCurrentDevice(device);
+    g_device = device;
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/tensor_impl.cpp:34]: do we need to sanity check here? (can someone set a non-lazy device type?)

File: lazy_tensor_core/lazy_tensor_core/csrc/device.cpp [hunk 3]
```diff
@@ -4,20 +4,16 @@
 #include <c10/util/Optional.h>
 
 namespace torch_lazy_tensors {
-namespace {
-
-thread_local c10::optional<Device> g_current_device;
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/device.cpp:9]: it seems like you have moved the thread-local device from the bridge to the tensor_impl, but the basic semantics (one global setting per thread) remains the same.  I think that's fine, just confirming i understood correctly.

File: lazy_tensor_core/test/cpp/test_aten_ltc_ts_tensor.cpp [hunk 4]
```diff
@@ -1627,9 +1627,9 @@ TEST_F(AtenLtcTsTensorTest, TestNormInDimsKeep) {
 TEST_F(AtenLtcTsTensorTest, TestNormalTwoTensor) {
   at::Tensor mean = at::zeros({10, 10, 10}, at::dtype(at::kFloat));
   at::Tensor std = at::ones({10, 10, 10}, at::dtype(at::kFloat));
-  ForEachDevice([&](const Device& device) {
-    at::Tensor xla_mean = bridge::CreateLtcTensor(mean, device);
```
Comment (wconstab) [inline lazy_tensor_core/test/cpp/test_aten_ltc_ts_tensor.cpp:1631]: what's the reason for this change? (just curious)

File: lazy_tensor_core/lazy_tensor_core/csrc/aten_ltc_bridge.cpp [hunk 5]
```diff
@@ -247,16 +247,9 @@ c10::optional<Device> GetLtcDevice(const c10::optional<c10::Device>& device) {
...
-  if (ordinal < 0) {
-    c10::Device current_device = GetCurrentAtenDevice();
-    if (current_device.has_index()) {
-      ordinal = current_device.index();
-    }
-  }
-  if (ordinal < 0) {
-    return GetCurrentDevice();
-  }
+  // Ordinal doesn't make any sense now given
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/aten_ltc_bridge.cpp:250]: it's not strictly true, sometimes people doing single-gpu work on a shared machine still want to set ordinal to a particular gpu to avoid conflict.  I don't know if that even applies to lazy:{ordinal} though, or if we override the actual GPU ordinal inside the backend.  That's something we can fix later, but I just wanted you to be aware.

File: lazy_tensor_core/lazy_tensor_core/csrc/debug_util.cpp [hunk 6]
```diff
@@ -102,7 +102,7 @@ std::string DebugUtil::GetTensorsGraphInfo(c10::ArrayRef<LazyTensor> tensors,
     graph_str = ir::DumpUtil::ToDot(root_nodes);
   } else if (format == GraphFormat::kBackend) {
     graph_str = ir::DumpUtil::ToBackend(
-        root_values, unique_device ? *unique_device : GetCurrentDevice());
+        root_values, unique_device ? *unique_device : Device());
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/debug_util.cpp:105]: I think it's a little messy to replace a 'current device' function with calling the default device ctor.  Is your plan that later on, the Device() ctor will actually query the backend for an appropriate default?  or does Device() leave type_ uninitialized and the surrounding code is supposed to understand that to mean no setting has been set yet?

**Review by wconstab — COMMENTED** [review]
overall i'm pretty happy with getting device-related functions out of the bridge; it seemed like one extra layer that isn't strictly necessary, and you picked up on that.

I had some quesitons about the use of default 'Device()' ctor;  Maybe if you explain the reasoning to me i'm ok with it.

File: lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp (hunk 1 above)
Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp:57]: The default `Device` constructor should set the default device type according to the backend interface and set the ordinal to zero (which can be changed if later on we add back API to support default device).

Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp:57]: > Device constructor should set the default device type according to the backend interface 
is this something you planned to add later?  Or i missed it and it's already there?

File: lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp [hunk 7]
```diff
@@ -578,9 +567,21 @@ void InitLtcModuleBindings(py::module m) {
...
+    // ii) Setting ordinal like lazy:0 doesn't make any sense now as distributed
+    // training is still under development.
+    LOG(ERROR) << "Setting the default device is deprecated. Use "
+                  "_ltc_set_default_device_type to set the default "
+                  "device type instead.";
+    return;
+  });
+  m.def("_ltc_get_default_device", []() {
+    // It's always lazy:0 given distributed training is not supported yet.
+    return "lazy:0";
```
[the remark this answers is not in the document]
Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp:583]: I assume if pytorch/xla has fully adopted LTC, they would just use `lazy` device instead of `xla` device. Also, the previous logic won't return `xla:0` as well given the bridge will always use the lazy keyword: https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/lazy_tensor_core/csrc/aten_ltc_bridge.cpp#L263.

File: lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp (hunk 1 above)
Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp:57]: Added a TODO in the default ctor' impl. Holding that on your backend API change.

File: lazy_tensor_core/lazy_tensor_core/csrc/aten_ltc_bridge.cpp (hunk 5 above)
Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/aten_ltc_bridge.cpp:250]: Since [TSBackendImpl::GetLocalDevices()](https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/backend_impl.cpp#L178) only support one element, I don't think we can set "lazy:1" currently. Also see: [link](https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/lazy_tensor_core/csrc/aten_ltc_bridge.cpp#L34). Let me rephrase the comments.

File: lazy_tensor_core/lazy_tensor_core/csrc/debug_util.cpp (hunk 6 above)
Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/debug_util.cpp:105]: Yea, I added a TODO for the default ctor of the device to query the default device type or default device in the future. Let me add a comment in the default ctor to make it more clear.

File: lazy_tensor_core/lazy_tensor_core/csrc/device.cpp (hunk 3 above)
Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/device.cpp:9]: Yea, your are right!

File: lazy_tensor_core/lazy_tensor_core/csrc/tensor_impl.cpp (hunk 2 above)
Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/tensor_impl.cpp:34]: Let me add a debug assert. It seems unlikely.

File: lazy_tensor_core/test/cpp/test_aten_ltc_ts_tensor.cpp (hunk 4 above)
Comment (alanwaketan) [reply lazy_tensor_core/test/cpp/test_aten_ltc_ts_tensor.cpp:1631]: Just to align the behaviors of these several tests with the remaining and to remove the additional overload of `ForEachDevice` that supports a backend device type. It's now all c10::Device.

File: lazy_tensor_core/lazy_tensor_core/csrc/aten_ltc_bridge.cpp (hunk 5 above)
Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/aten_ltc_bridge.cpp:250]: ok; yea this part we can fix later

File: lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp:583]: ok. this is probably TBD in terms of how fast xla makes the change, or if they consider keeping xla:: device.

But I still prefer to merge as-is and then revisit this once we're focusing more on e2e design of the UI and pybindings.

**Review by wconstab — APPROVED** [review]
ok, lets merge it. I think there are some things we'll have to change again later, but it makes sense to leave them for now since we haven't done an e2e design for all of these areas yet.

File: lazy_tensor_core/lazy_tensor_core/csrc/aten_ltc_bridge.cpp (hunk 5 above)
Comment (JackCaoG) [reply lazy_tensor_core/lazy_tensor_core/csrc/aten_ltc_bridge.cpp:250]: This means we hardcode the `lazy:0` to be the default right? I felt like it is better to call `GetDefaultDevice` here instead. 

In the case of pt/lxa, `xla;0` is usually a `cpu` and `xla:1` is a `tpu`. We don't want to return `xla:0` as the default.

Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/aten_ltc_bridge.cpp:250]: Okay, then it looks like we need a way to actually query the default device in the backend interface. W.R.T this particular code, it's going to remove by Will in #67869 anyway. So it shouldn't matter.

File: lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp:583]: Let me also add a TODO here suggest that it's deprecated like the above `_ltc_set_default_device`.

File: lazy_tensor_core/lazy_tensor_core/csrc/aten_ltc_bridge.cpp (hunk 5 above)
Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/aten_ltc_bridge.cpp:250]: interesting. so when users use xla for single node tpu today, they have to move their tensors to xla:1?  

(or, they move to `xla` and the default device ordinal is 1?)

Comment (JackCaoG) [reply lazy_tensor_core/lazy_tensor_core/csrc/aten_ltc_bridge.cpp:250]: We don't ask user to manually set device name like `xla:1`. We have a API [xm.xla_device()](https://github.com/pytorch/xla/blob/f9eb88c7216332dd11600e58ffd3067622c263fd/torch_xla/core/xla_model.py#L218) which will also set the [defualt_device](https://github.com/pytorch/xla/blob/f9eb88c7216332dd11600e58ffd3067622c263fd/torch_xla/core/xla_model.py#L240).

**Review by JackCaoG — APPROVED** [review]
Only concern is that `Device()` returns `lazy:0`. If we can make it return the default device (from the backend), this change lgtm.

**Comment by alanwaketan** [comment]
> Only concern is that `Device()` returns `lazy:0`. If we can make it return the default device (from the backend), this change lgtm.

Yea, definitely. Thanks for approving this pull request.

**Comment by alanwaketan** [comment]
> ok, lets merge it. I think there are some things we'll have to change again later, but it makes sense to leave them for now since we haven't done an e2e design for all of these areas yet.

Yea, definitely needs more e2e designs on those areas in the future. Thanks for approving the pull request.
